pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -216,22 +216,49 @@ func (e *Controller) addPod(obj interface{}) {
 
 func podToEndpointAddressForService(svc *v1.Service, pod *v1.Pod) (*v1.EndpointAddress, error) {
 	var endpointIP string
+	ipFamily := v1.IPv4Protocol
 
 	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
 		// In a legacy cluster, the pod IP is guaranteed to be usable
 		endpointIP = pod.Status.PodIP
 	} else {
-		ipv6Service := endpointutil.IsIPv6Service(svc)
+		//feature flag enabled and pods may have multiple IPs
+		if len(svc.Spec.IPFamilies) > 0 {
+			// controller is connected to an api-server that correctly sets IPFamilies
+			ipFamily = svc.Spec.IPFamilies[0] // this works for headful and headless
+		} else {
+			// controller is connected to an api server that does not correctly
+			// set IPFamilies (e.g. old api-server during an upgrade)
```

Comment:
I have been arguing with @thockin about a clear messaging around upgrade process (api-server => control-plane=>nodes). However in my mind i think it is best to handle the case even if we don't exactly support it. Instead of breaking the entire cluster (in indeterministic ways panic, loss of traffic etc) we gracefully handle it, and we log the case.